lib/azure/armrest/configuration.rb

```diff
@@ -86,39 +83,51 @@ def clear_caches
       def initialize(args)
         # Use defaults, and override with provided arguments
         options = {
-          :api_version      => '2015-01-01',
-          :accept           => 'application/json',
-          :content_type     => 'application/json',
-          :grant_type       => 'client_credentials',
-          :token_expiration => Time.new(0).utc,
-          :proxy            => ENV['http_proxy'],
-          :ssl_version      => 'TLSv1',
-        }.merge(args)
+          :api_version  => '2015-01-01',
+          :accept       => 'application/json',
+          :content_type => 'application/json',
+          :grant_type   => 'client_credentials',
+          :proxy        => ENV['http_proxy'],
+          :ssl_version  => 'TLSv1',
+        }.merge(args.symbolize_keys)
+
+        user_token = options.delete(:token)
+        user_token_expiration = options.delete(:token_expiration)
 
         options.each { |key, value| send("#{key}=", value) }
 
         unless client_id && client_key && tenant_id && subscription_id
           raise ArgumentError, "client_id, client_key, tenant_id and subscription_id must all be specified"
         end
 
+        if user_token && user_token_expiration
+          set_token(user_token, user_token_expiration)
+        elsif user_token || user_token_expiration
+          raise ArgumentError, "token and token_expiration must be both specified"
+        end
+
         # Allows for URI objects or Strings.
         @proxy = @proxy.to_s if @proxy
 
-        # A one-time lookup table set at the class level
         @providers = fetch_providers
-        set_provider_info(@providers) if self.class.provider_version_cache.empty?
+        set_provider_api_versions
+      end
+
+      def hash
+        [tenant_id, client_id, client_key].join('_').hash
+      end
+
```

Comment:
`hash` and `eql?` are twin overridden methods from `Object`, similar to `hashCode` and `equals` in Java.